File: dotnet/Lexy/Lexy.Compiler/Language/Types/PrimitiveType.cs
using Lexy.Compiler.Language.Tables;

namespace Lexy.Compiler.Language.Types
{
    public class PrimitiveType : VariableType
    {
        public static PrimitiveType Boolean => new PrimitiveType(TypeNames.Boolean);
        public static PrimitiveType String => new PrimitiveType(TypeNames.String);
        public static PrimitiveType Number => new PrimitiveType(TypeNames.Number);
        public static PrimitiveType Date => new PrimitiveType(TypeNames.Date);

        public string Type { get; }

        public PrimitiveType(string type)
        {
            Type = type;
        }

        protected bool Equals(PrimitiveType other)
        {
            return Type == other.Type;
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != this.GetType()) return false;
            return Equals((PrimitiveType)obj);
        }

        public override int GetHashCode()
        {
            return (Type != null ? Type.GetHashCode() : 0);
        }

        public override string ToString() => Type;
    }
}

File: typescipt/src/language/types/primitiveType.ts


export class PrimitiveType extends VariableType {
   public static PrimitiveType Boolean =new(): >;
   public static PrimitiveType String =new(): >;
   public static PrimitiveType Number =new(): >;
   public static PrimitiveType Date =new(): >;

   public string Type

   constructor(type: string) {
     Type = type;
   }

   protected equals(other: PrimitiveType): boolean {
     return Type == other.Type;
   }

   public override equals(obj: object): boolean {
     if (ReferenceEquals(null, obj)) return false;
     if (ReferenceEquals(this, obj)) return true;
     if (obj.GetType() != GetType()) return false;
     return Equals((PrimitiveType)obj);
   }

   public override getHashCode(): number {
     return Type != null ? Type.GetHashCode() : 0;
   }

   public override toString(): string {
     return Type;
   }
}
